packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -26,41 +27,112 @@ class Plugin {
   Plugin({
     this.name,
     this.path,
-    this.androidPackage,
-    this.iosPrefix,
-    this.macosPrefix,
-    this.pluginClass,
+    this.platformPlugins,
   });
 
+  /// Parses [Plugin] specification from the provided pluginYaml.
+  ///
+  /// This currently supports two formats. V1 (Legacy) and V2 (Multi-platform).
+  /// Example V1 format. Which is deprecated.
+  /// flutter:
+  ///  plugin:
+  ///    androidPackage: io.flutter.plugins.sample
+  ///    iosPrefix: FLT
+  ///    macosPrefix: FLT
+  ///    pluginClass: SamplePlugin
+  ///
+  /// Example V2 format.
+  /// flutter:
+  ///  plugin:
+  ///    platforms:
+  ///      android:
+  ///        package: io.flutter.plugins.sample
+  ///        pluginClass: SamplePlugin
+  ///      ios:
+  ///        classPrefix: FLT
+  ///        pluginClass: SamplePlugin
+  ///      macos:
+  ///        classPrefix: FLT
+  ///        pluginClass: SamplePlugin
   factory Plugin.fromYaml(String name, String path, dynamic pluginYaml) {
-    String androidPackage;
-    String iosPrefix;
-    String macosPrefix;
-    String pluginClass;
-    if (pluginYaml != null) {
-      androidPackage = pluginYaml['androidPackage'];
-      iosPrefix = pluginYaml['iosPrefix'] ?? '';
+    // TODO(kaushikiska): Fail if both old and new formats are present.
+    if (pluginYaml != null && pluginYaml['platforms'] != null) {
+      return Plugin._fromMultiPlatformYaml(name, path, pluginYaml);
+    } else {
+      return Plugin._fromLegacyYaml(name, path, pluginYaml); // ignore: deprecated_member_use_from_same_package
+    }
+  }
+
+  factory Plugin._fromMultiPlatformYaml(String name, String path, dynamic pluginYaml) {
+    assert (pluginYaml != null && pluginYaml['platforms'] != null,
+            'Invalid multi-platform plugin specification.');
+    final dynamic platforms = pluginYaml['platforms'];
+
+    final Map<String, PlatformPlugin> platformPlugins = <String, PlatformPlugin>{};
+
+    if (platforms[AndroidPlugin.kConfigKey] != null) {
+      platformPlugins[AndroidPlugin.kConfigKey] =
+          AndroidPlugin.fromYaml(name, platforms[AndroidPlugin.kConfigKey]);
+    }
+
+    if (platforms[IOSPlugin.kConfigKey] != null) {
+      platformPlugins[IOSPlugin.kConfigKey] =
+          IOSPlugin.fromYaml(name, platforms[IOSPlugin.kConfigKey]);
+    }
+
+    if (platforms[MacOSPlugin.kConfigKey] != null) {
+      platformPlugins[MacOSPlugin.kConfigKey] =
+          MacOSPlugin.fromYaml(name, platforms[MacOSPlugin.kConfigKey]);
+    }
+
```

Comment:
I would restructure this so that we can grab unspecified platforms as well as known platforms.